cmd/intermediary-auditor/main.go

```diff
@@ -0,0 +1,191 @@
+package main
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/csv"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"os"
+	"strings"
+	"sync"
+	"time"
+)
+
+const (
+	r3 = "R3"
+	x3 = "Let's Encrypt Authority X3"
+)
+
+var debugMode bool
+var statsCsv string
+var wg sync.WaitGroup
+
+// chainContainsR3 checks if a chain of certs contains a certificate
+// where the Subject Common Name matches the const of r3
+func chainContainsR3(chain []*x509.Certificate) bool {
+	for _, cert := range chain[1:] {
+		if cert.Subject.CommonName == r3 {
+			return true
+		}
+	}
+	return false
+}
+
+// chainContainsX3 checks if a chain of certs contains a certificate
+// where the Subject Common Name matches the const of x3
+func chainContainsX3(chain []*x509.Certificate) bool {
+	for _, cert := range chain[1:] {
+		if cert.Subject.CommonName == x3 {
+			return true
+		}
+	}
+	return false
+}
+
+// rawToChain marshals a slice of byte slices representing an x.509
+// certificate chain to a slice of *x.509Certificate objects
+func rawToChain(rawCerts [][]byte) []*x509.Certificate {
+	chain := []*x509.Certificate{}
+	for _, rawCert := range rawCerts {
+		cert, err := x509.ParseCertificate(rawCert)
+		if err != nil {
+			continue
+		}
+		chain = append(chain, cert)
+	}
+	return chain
+}
+
+// chaing2String is used solely if debug is true. Iterates from the
+// leaf (end-entity) certificate all the way up the chain building a
+// string to represent the Subject Common Name and Issuer Common Name
+// for each Certificate
+func chaing2String(chain []*x509.Certificate) string {
+	var sb strings.Builder
+	sb.WriteString(fmt.Sprintf("leafCert: [subjectCN: %s | issuerCN: %s]", chain[0].Subject.CommonName, chain[0].Issuer.CommonName))
+	for num, cert := range chain[1:] {
+		sb.WriteString(fmt.Sprintf(" -> chainCert%d: [subjectCN: %s | issuerCN: %s] ", num, cert.Subject.CommonName, cert.Issuer.CommonName))
+	}
+	return sb.String()
+}
+
+// auditChain for a given slice of byte slices representing an x.509
+// certificate chain, if the Issuer Common Name is const x3 or r3,
+// validates that the resulting chain of x509 Certificates contains the
+// corresponding x3 or r3 intermediate that issued the leaf Certificate.
+// If a mis-match is present, a string containing the Subject Common
+// Name of the leaf certificate is returned, else, in all other cases an
+// empty string is returned.
+func auditChain(rawCerts [][]byte) string {
+	chain := rawToChain(rawCerts)
+	leafIssuerCN := chain[0].Issuer.CommonName
+	if len(chain) > 1 {
+		if debugMode == true {
+			fmt.Println(chaing2String(chain))
+		}
+		if leafIssuerCN != r3 && leafIssuerCN != x3 {
+			return ""
+		}
+		if leafIssuerCN == r3 && chainContainsR3(chain) {
+			return ""
+		}
+		if leafIssuerCN == x3 && chainContainsX3(chain) {
```

Comment:
I think for this line we should just say `if leafIssuerCN == x3`. The reasoning is this: We want this script to catch _new_ problems only. Plenty of people have the problem that they're deploying an X3-signed leaf without a chain. We don't care about those cases for this particular task. We may later want to adapt this to catch that category of cases, though.

```suggestion
		if leafIssuerCN == x3 {
```

Reply:
Awesome, and this drastically simplifies the function.